src/download.rs

```diff
@@ -18,47 +22,130 @@ struct LatestReleaseApiResponse {
     name: String,
 }
 
-pub fn forc_bin_tarball_name() -> Result<String> {
-    let os = match std::env::consts::OS {
-        "macos" => "darwin",
-        "linux" => "linux",
-        unsupported_os => bail!("Unsupported os: {}", unsupported_os),
-    };
-    let architecture = match std::env::consts::ARCH {
-        "aarch64" => "arm64",
-        "x86_64" => "amd64",
-        unsupported_arch => bail!("Unsupported architecture: {}", unsupported_arch),
-    };
-
-    Ok(format!("forc-binaries-{}_{}.tar.gz", os, architecture))
+#[derive(Debug, PartialEq, Eq)]
+pub struct DownloadCfg {
+    pub name: String,
+    pub version: String,
+    release_url: String,
 }
 
-pub fn fuel_core_bin_tarball_name(version: &str) -> Result<String> {
-    let architecture = match std::env::consts::ARCH {
-        "aarch64" => "aarch64",
-        "x86_64" => "x86_64",
-        unsupported_arch => bail!("Unsupported architecture: {}", unsupported_arch),
-    };
-
-    let vendor = match std::env::consts::OS {
-        "macos" => "apple",
-        _ => "unknown",
-    };
+impl DownloadCfg {
+    pub fn new(name: &str, version: Option<String>) -> Result<DownloadCfg> {
+        Ok(Self {
+            name: name.to_string(),
+            version: match version {
+                Some(version) => {
+                    if version.starts_with('v') {
+                        version
+                    } else {
+                        "v".to_string() + &version
+                    }
+                }
+                None => {
+                    let latest_tag_url = match name {
+                        "forc" => format!(
+                            "{}{}/{}",
+                            GITHUB_API_REPOS_BASE_URL, SWAY_REPO, RELEASES_LATEST
+                        ),
+                        "fuel-core" => format!(
+                            "{}{}/{}",
+                            GITHUB_API_REPOS_BASE_URL, FUEL_CORE_REPO, RELEASES_LATEST
+                        ),
+                        "fuelup" => format!(
+                            "{}{}/{}",
+                            GITHUB_API_REPOS_BASE_URL, FUELUP_REPO, RELEASES_LATEST
+                        ),
+                        _ => bail!("Unrecognized component: {}", name),
+                    };
+                    if let Ok(result) = get_latest_tag(&latest_tag_url) {
+                        result
+                    } else {
+                        bail!("Error getting latest tag for component: {}", name);
+                    }
+                }
+            },
+            release_url: match name {
+                "forc" => SWAY_RELEASE_DOWNLOAD_URL.to_string(),
+                "fuel-core" => FUEL_CORE_RELEASE_DOWNLOAD_URL.to_string(),
+                "fuelup" => FUELUP_RELEASE_DOWNLOAD_URL.to_string(),
```

Comment:
I think it might be worth moving these component names into constants as they're used in quite a few places and it might help to avoid typos, e.g.

```rust
pub mod component {
    pub const FORC: &str = "forc";
    pub const FUEL_CORE: &str = "fuel-core";
    pub const FUELUP: &str = "fuelup";
}

// ...

match name {
    component::FORC => SWAY_RELEASE_DOWNLOAD_URL.to_string(),
// etc
```